File: sample-chat-obj-c/sample-chat/Components/ChatScreen/Views/InputToolbar/InputToolbar.m
//
//  InputToolbar.m
//  samplechat
//
//  Created by Injoit on 2/25/19.
//  Copyright © 2019 Quickblox. All rights reserved.
//

#import "InputToolbar.h"
#import "UIView+Chat.h"
#import "ChatResources.h"


static void * kInputToolbarKeyValueObservingContext = &kInputToolbarKeyValueObservingContext;

@interface InputToolbar()

@property (assign, nonatomic) BOOL isObserving;
@property (strong, nonatomic) UIButton *sendButton;

@end

@implementation InputToolbar
@dynamic delegate;

#pragma mark - Initialization

- (void)awakeFromNib {
    [super awakeFromNib];
    
    [self commonInit];
}

- (instancetype)init {
    self = [super init];
    if (self) {
        [self commonInit];
    }
    return self;
}

- (void)commonInit {
    [self setTranslatesAutoresizingMaskIntoConstraints:NO];
    [self layoutIfNeeded];
    self.isObserving = NO;
    self.sendButtonOnRight = YES;
    self.preferredDefaultHeight = 44.0f;
    ToolbarContentView *toolbarContentView = [self loadToolbarContentView];
    [self addSubview:toolbarContentView];
    _contentView = toolbarContentView;
    [self setShadowImage:UIImage.new forToolbarPosition:UIBarPositionAny];
    [self pinAllEdgesOfSubview:toolbarContentView];
    [self setNeedsUpdateConstraints];
    
    [self addObservers];
    [self toggleSendButtonEnabledIsUploaded:NO];
}


- (ToolbarContentView *)loadToolbarContentView {
    
    NSArray *nibViews = [[ChatResources resourceBundle] loadNibNamed:NSStringFromClass([ToolbarContentView class])
                                                                 owner:nil
                                                               options:nil];
    return nibViews.firstObject;
}

- (void)dealloc {
    [self removeObservers];
    [self.contentView removeFromSuperview];
    _contentView = nil;
}

#pragma mark - Setters

- (void)setPreferredDefaultHeight:(CGFloat)preferredDefaultHeight {
    NSParameterAssert(preferredDefaultHeight > 0.0f);
    _preferredDefaultHeight = preferredDefaultHeight;
}

#pragma mark - Actions

- (void)leftBarButtonPressed:(UIButton *)sender {
    [self.delegate messagesInputToolbar:self didPressLeftBarButton:sender];
}

- (void)rightBarButtonPressed:(UIButton *)sender {
    [sender setEnabled:NO];
    [self.delegate messagesInputToolbar:self didPressRightBarButton:sender];
}

#pragma mark - Input toolbar

- (void)toggleButtons {
    
    BOOL hasText = self.contentView.textView.text.length > 0;
    BOOL hasTextAttachment = [self.contentView.textView hasTextAttachment];
    BOOL hasDataToSend = hasText || hasTextAttachment;
    
    UIButton *buttonToUpdate;
    UIView *buttonContainer;
    if (self.sendButtonOnRight) {
        buttonToUpdate = self.contentView.rightBarButtonItem;
        buttonContainer = self.contentView.rightBarButtonContainerView;
    } else {
        buttonToUpdate = self.contentView.leftBarButtonItem;
        buttonContainer = self.contentView.leftBarButtonContainerView;
    }
    
    buttonToUpdate.hidden = !hasDataToSend;
    buttonToUpdate.enabled = [self.contentView.textView hasText];
}

- (void)addCenterConstraintsToItem:(UIView *)itemToAdd {

}

- (void)setupBarButtonEnabledLeft:(Boolean)left andRight:(Boolean)right {
    self.contentView.rightBarButtonItem.enabled = right;
    self.contentView.leftBarButtonItem.enabled = left;
}

- (void)toggleSendButtonEnabledIsUploaded:(BOOL)isUploaded {
    BOOL hasText = [self.contentView.textView hasText];
    if (self.sendButtonOnRight || isUploaded) {
        self.contentView.rightBarButtonItem.enabled = hasText || isUploaded;
    }
}

#pragma mark - Key-value observing

- (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(NSDictionary *)change context:(void *)context {
    if (context == kInputToolbarKeyValueObservingContext) {
        
        if (object == self.contentView) {
            
            if ([keyPath isEqualToString:NSStringFromSelector(@selector(leftBarButtonItem))]) {
                
                [self.contentView.leftBarButtonItem removeTarget:self
                                                          action:NULL
                                                forControlEvents:UIControlEventTouchUpInside];
                
                [self.contentView.leftBarButtonItem addTarget:self
                                                       action:@selector(leftBarButtonPressed:)
                                             forControlEvents:UIControlEventTouchUpInside];
            }
            else if ([keyPath isEqualToString:NSStringFromSelector(@selector(rightBarButtonItem))]) {
                
                [self.contentView.rightBarButtonItem removeTarget:self
                                                           action:NULL
                                                 forControlEvents:UIControlEventTouchUpInside];
                
                [self.contentView.rightBarButtonItem addTarget:self
                                                        action:@selector(rightBarButtonPressed:)
                                              forControlEvents:UIControlEventTouchUpInside];
            }
            
            [self toggleSendButtonEnabledIsUploaded:NO];
        }
    }
}

- (void)addObservers {
    
    if (self.isObserving) {
        return;
    }
    
    [self.contentView addObserver:self
                       forKeyPath:NSStringFromSelector(@selector(leftBarButtonItem))
                          options:0
                          context:kInputToolbarKeyValueObservingContext];
    
    [self.contentView addObserver:self
                       forKeyPath:NSStringFromSelector(@selector(rightBarButtonItem))
                          options:0
                          context:kInputToolbarKeyValueObservingContext];
    
    self.isObserving = YES;
}

- (void)removeObservers {
    
    if (!self.isObserving) {
        return;
    }
    
    @try {
        [_contentView removeObserver:self
                          forKeyPath:NSStringFromSelector(@selector(leftBarButtonItem))
                             context:kInputToolbarKeyValueObservingContext];
        
        [_contentView removeObserver:self
                          forKeyPath:NSStringFromSelector(@selector(rightBarButtonItem))
                             context:kInputToolbarKeyValueObservingContext];
    }
    @catch (NSException *__unused exception) { }
    
    self.isObserving = NO;
}
@end

File: sample-chat-swift/sample-chat-swift/Components/ChatScreen/Chat/CustomUI/InputToolBar/InputToolbar.swift
//
//  InputToolbar.swift
//  sample-chat-swift
//
//  Created by Injoit on 1/28/19.
//  Copyright © 2019 Quickblox. All rights reserved.
//

import UIKit

/**
 *  The `InputToolbarDelegate` protocol defines methods for interacting with
 *  a `ChatMessageInputToolbar` object.
 */
protocol InputToolbarDelegate: UIToolbarDelegate {
    /**
     *  Tells the delegate that the toolbar's `rightBarButtonItem` has been pressed.
     *
     *  @param toolbar The object representing the toolbar sending this information.
     *  @param sender  The button that received the touch event.
     */
    func messagesInputToolbar(_ toolbar: InputToolbar, didPressRightBarButton sender: UIButton)
    /**
     *  Tells the delegate that the toolbar's `leftBarButtonItem` has been pressed.
     *
     *  @param toolbar The object representing the toolbar sending this information.
     *  @param sender  The button that received the touch event.
     */
    func messagesInputToolbar(_ toolbar: InputToolbar, didPressLeftBarButton sender: UIButton)
}

class InputToolbar: UIToolbar {
    
    private var rightButtonStatusObserver:NSKeyValueObservation?
    private var leftButtonStatusObserver:NSKeyValueObservation?
    
    /**
     *  An instance of `InputToolbar` defines the input toolbar for
     *  composing a new message. It is displayed above and follow the movement of
     *  the system keyboard.
     */
    
    /**
     *  The object that acts as the delegate of the toolbar.
     */
    weak var inputToolbarDelegate: InputToolbarDelegate?
    
    override weak var delegate: UIToolbarDelegate? {
        didSet{
            inputToolbarDelegate = delegate as? InputToolbarDelegate
        }
    }
    /**
     *  Returns the content view of the toolbar. This view contains all subviews of the toolbar.
     */
    lazy public private(set) var contentView = loadToolbarContentView()
    /**
     *  A boolean value indicating whether the send button is on the right side of the toolbar or not.
     *
     *  @discussion The default value is `YES`, which indicates that the send button is the right-most subview of
     *  the toolbar's `contentView`. Set to `NO` to specify that the send button is on the left. This
     *  property is used to determine which touch events correspond to which actions.
     *
     *  @warning Note, this property *does not* change the positions of buttons in the toolbar's content view.
     *  It only specifies whether the `rightBarButtonItem `or the `leftBarButtonItem` is the send button.
     *  The other button then acts as the accessory button.
     */
    var sendButtonOnRight = true
    /**
     *  Specifies the default height for the toolbar. The default value is `44.0`. This value must be positive.
     */
    var preferredDefaultHeight: CGFloat = 44.0 {
        didSet {
            assert(preferredDefaultHeight > 0.0, "Invalid parameter not satisfying: preferredDefaultHeight > 0.0")
        }
    }
    
    /**
     *  Enables or disables the send button based on whether or not its `textView` has text.
     *  That is, the send button will be enabled if there is text in the `textView`, and disabled otherwise.
     */
    func setupBarButtonsEnabled(left: Bool, right: Bool) {
        contentView.rightBarButtonItem?.isEnabled = right
        contentView.leftBarButtonItem?.isEnabled = left
    }
    
    func toggleSendButtonEnabled(isUploaded: Bool) {
        let hasText = contentView.textView.hasText
        if sendButtonOnRight == true || isUploaded == true {
            contentView.rightBarButtonItem?.isEnabled = hasText || isUploaded
        }
    }
    
    /**
     *  Loads the content view for the toolbar.
     *
     *  @discussion Override this method to provide a custom content view for the toolbar.
     *
     *  @return An initialized `ToolbarContentView` if successful, otherwise `nil`.
     */
    open func loadToolbarContentView() -> ToolbarContentView {
        let nibName = String(describing:ToolbarContentView.self)
        let objects = Bundle.main.loadNibNamed(nibName, owner: nil)
        let toolbarContentView = objects!.first as! ToolbarContentView
        return toolbarContentView
    }
    
    // MARK: - Initialization
    override func awakeFromNib() {
        super.awakeFromNib()
        commonInit()
    }
    
    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
    }
    
    func commonInit() {
        translatesAutoresizingMaskIntoConstraints = false
        layoutIfNeeded()
        let toolbarContentView: ToolbarContentView? = loadToolbarContentView()
        
        if let toolbarContentView = toolbarContentView {
            addSubview(toolbarContentView)
            contentView = toolbarContentView
            setShadowImage(UIImage(), forToolbarPosition: UIBarPosition.any)
        }
        pinAllEdges(ofSubview: toolbarContentView)
        setNeedsUpdateConstraints()
        
        addObservers()
        toggleSendButtonEnabled(isUploaded: false)
    }
    
    deinit {
        removeObservers()
        contentView.removeFromSuperview()
    }
    
    // MARK: - Actions
    @objc func leftBarButtonPressed(_ sender: UIButton) {
        inputToolbarDelegate?.messagesInputToolbar(self, didPressLeftBarButton: sender)
    }
    
    @objc func rightBarButtonPressed(_ sender: UIButton) {
        sender.isEnabled = false
        inputToolbarDelegate?.messagesInputToolbar(self, didPressRightBarButton: sender)
    }
    
    // MARK: - observing
    func addObservers() {
        leftButtonStatusObserver = contentView.observe(\ToolbarContentView.leftBarButtonItem,
                                                       options: [.new, .old],
                                                       changeHandler: {[weak self] (contentView, change) in
                                                        guard let self = self else {
                                                            return
                                                        }
                                                        
                                                        guard let leftButton = change.newValue  as? UIButton else {
                                                            return
                                                        }
                                                        leftButton.addTarget(self,
                                                                             action: #selector(self.leftBarButtonPressed(_:)),
                                                                             for: .touchUpInside)
        })
        
        rightButtonStatusObserver = contentView.observe(\ToolbarContentView.rightBarButtonItem,
                                                        options: [.new, .old],
                                                        changeHandler: {[weak self] (contentView, change) in
                                                            guard let self = self else {
                                                                return
                                                            }
                                                            
                                                            guard let leftButton = change.newValue  as? UIButton else {
                                                                return
                                                            }
                                                            
                                                            leftButton.addTarget(self,
                                                                                 action: #selector(self.rightBarButtonPressed(_:)),
                                                                                 for: .touchUpInside)
        })
        
    }
    
    func removeObservers() {
        if let leftButtonObserver = leftButtonStatusObserver {
            leftButtonObserver.invalidate()
            leftButtonStatusObserver = nil
        }
        if let rightButtonObserver = rightButtonStatusObserver {
            rightButtonObserver.invalidate()
            rightButtonStatusObserver = nil
        }
    }
}
